### PR TITLE
added file format backward compatability tests

### PR DESCRIPTION
**Pull request by cseed**
This adds about 800KB total to the repo.

I added a notion of supportedCodec, which includes all the codecs except DirectCodec (which currently depends on the in memory representation of region values which we don't want to promise).

FYI @tpoterba I also added a micro optimization: if you get a field of a struct declaration, just grab out the AST for the field.  This fixed some massive expression size blowup in create_all_values_datasets.

### CODE REVIEW

**Comment by tpoterba** [comment]
can you add the rest of the types that are missing? tuple, some numerics, bool?

**Comment by cseed** [comment]
added tuple, there were already bool, numerics, etc. (although not all at the top level).  I felt like it was fine as long as they were in there, somewhere.

File: python/hail/tests.py [hunk 1]
```diff
@@ -37,6 +37,88 @@ def convert_struct_to_dict(x):
     else:
         return x
 
+def create_all_values_datasets():
+    all_values = hl.struct(
+        f32=hl.float32(3.14),
+        i64=hl.int64(-9),
+        m=hl.null(hl.tfloat64),
+        astruct=hl.struct(a=5, b=5.5),
```
Comment (jigold) [inline python/hail/tests.py]: do we need empty structs, sets, tuples, etc?

File: python/hail/tests.py [hunk 2]
```diff
@@ -37,6 +37,88 @@ def convert_struct_to_dict(x):
...
+
+    matrix_table_dir = 'backward_compatability/{}/matrix_table'.format(file_version)
+    if not os.path.exists(matrix_table_dir):
+        os.makedirs(matrix_table_dir)
+
+    i = 0
+    for codec in supported_codecs:
+        all_values_table.write('{}/{}.ht'.format(table_dir, i), overwrite=True, _codec_spec=codec.toString())
+        all_values_matrix_table.write('{}/{}.hmt'.format(matrix_table_dir, i), overwrite=True, _codec_spec=codec.toString())
+        i = i + 1
```
Comment (jigold) [inline python/hail/tests.py]: i += 1

File: python/hail/tests.py [hunk 3]
```diff
@@ -37,6 +37,88 @@ def convert_struct_to_dict(x):
...
+    all_values_matrix_table = (hl.utils.range_matrix_table(3, 2, n_partitions=2)
+                               .annotate_globals(**prefix(all_values, 'global_'))
+                               .annotate_rows(**prefix(all_values, 'row_'))
+                               .annotate_cols(**prefix(all_values, 'col_'))
+                               .annotate_entries(**prefix(all_values, 'entry_'))
+                               .cache())
+    return all_values_table, all_values_matrix_table
+
+def create_backward_compatability_test_files():
+    import os
```
Comment (jigold) [inline python/hail/tests.py]: Is this just run when needed? I don't see an actual call to this function.

File: python/hail/tests.py [hunk 4]
```diff
@@ -37,6 +37,88 @@ def convert_struct_to_dict(x):
...
+        table_dir = test_file('backward_compatability/1.0.0/table')
+        matrix_table_dir = test_file('backward_compatability/1.0.0/matrix_table')
+
+        n = 0
+        i = 0
+        f = os.path.join(table_dir, '{}.ht'.format(i))
+        while os.path.exists(f):
+            ds = hl.read_table(f)
+            self.assertTrue(ds._same(all_values_table))
+            i = i + 1
```
Comment (jigold) [inline python/hail/tests.py]: += here and below

File: python/hail/tests.py (hunk 1 above)
Comment (cseed) [reply python/hail/tests.py]: I added a variety of missing aggregate types.  I didn't really mean to make this exhaustive.  (We can always go back to this version and add more stuff to the files.)

File: python/hail/tests.py (hunk 3 above)
Comment (cseed) [reply python/hail/tests.py]: Correct, I ran it once in this version to produce the files in `src/test/resources/backward_compatability`.

**Comment by cseed** [comment]
I'm not really sure, either.  I figured as long as every type was in there somewhere at least once and some things were missing, it would be fine.  This is just to check backward compatibility, so it's hard to imagine that (1) the current version passes all the tests, (2) it can load a reasonable collections of types and values from an old version, but (3) only, say, missing calls are broken.

**Pull request merged**